### PR TITLE
Optimize spark split enumeration

### PR DESCRIPTION
**Pull request by arhimondr**
When enumerating splits for large tables holding all the splits inmemory
can potentially be very expensive. In our deployment a single HiveSplit
takes 1.5kb of memory. However multiple splits are extremely
compressible. This patch introduces iterative split enumeration. Splits
are loaded and serialized+compressed in batches.

Test plan:

- Implement and run extensive unit tests to make sure split enumeration logic is not broken when enumerating in batches
- Add an in memory integration test with a batch size set to `1`, `2` and `4` splits to make sure the code path is exercised
- Run problematic production query that used to trigger driver OOM to make sure the patch resolves the problem
- Run standard verifier suite with standard settings as well as with altered batch size to `1`, `2` and `5` splits

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Review by viczhang861 — COMMENTED** [review]
Typo in commit message: compressible, inmemory

File: presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java [hunk 1]
```diff
@@ -322,36 +320,72 @@ private PrestoSparkTaskSourceRdd createTaskSourcesRdd(
...
-            long sequenceId = 0;
-            while (!splitSource.isFinished()) {
-                List<Split> splitBatch = getFutureValue(splitSource.getNextBatch(NOT_PARTITIONED, Lifespan.taskWide(), 1000)).getSplits();
-                for (Split split : splitBatch) {
-                    splits.add(new ScheduledSplit(sequenceId++, tableScan.getId(), split));
-                }
+        ListMultimap<Integer, SerializedPrestoSparkTaskSource> result = ArrayListMultimap.create();
+        AtomicInteger sequenceId = new AtomicInteger();
+        int splitAssignmentBatchSize = getSplitAssignmentBatchSize(session);
+        Int2LongMap partitionToAssignedSplitSizeMap = new Int2LongArrayMap();
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java:332]: Actually yes. Let me refactor the split assignment algorithms into a standalone, statefull class

**Comment by arhimondr** [comment]
Extracted split assignment logic into separate components. Thanks for the suggestion @viczhang861

File: presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkConfig.java [hunk 2]
```diff
@@ -36,6 +36,7 @@
     private String storageBasedBroadcastJoinStorage = "local";
     private DataSize sparkBroadcastJoinMaxMemoryOverride;
     private boolean smileSerializationEnabled = true;
+    private int splitAssignmentBatchSize = 1_000_000;
```
Comment (pgupta2) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkConfig.java:39]: Wondering if this batch size is too big?  As you mentioned in the summary, a single HiveSplit can take 1.5kb of memory. With this batch size, we are looking at 1.5gb of memory consumption (worst case), before we serialize and compress the splits. What are the cons of making this smaller to further reduce memory usage?

Comment (arhimondr) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkConfig.java:39]: This is a little bit of a tradeoff. A 1_000_000 splits is just a 100 splits per partition (assuming the maximum number of partitions is 10000). Getting to lower values might reduce compression efficiency.

**Review by viczhang861 — APPROVED** [review]
Much cleaner after creating the interface.

**Review by pgupta2 — APPROVED** [review]

**Pull request merged**